### ISSUE TITLE
await permissions.RequestPermissionsAsync simply not returning anything

### ISSUE THREAD

**Issue by remingtonsutton** [jamesmontemagno/PermissionsPlugin#50]
If you are creating an issue for a BUG please fill out this information. If you are asking a question or requesting a feature you can delete the sections below. 

**Failure to fill out this information will result in this issue being closed.** If you post a full stack trace in a bug it will be closed, please post it to http://gist.github.com and then post the link here.

## Bug Information

Version Number of Plugin: 1.2.1
Device Tested On:
Simulator Tested On:
Version of VS: Professional 2017
Version of Xamarin: 2.3.4.247
Versions of other things you are using: Target and build in api 25 on adroid

### Steps to reproduce the Behavior

var results = await permissions.RequestPermissionsAsync(Permission.Location);
status = results[Permission.Location];

### Expected Behavior
To return permission results.  Are there any conflicts with API 25?? I have included the overridden method in MainActivity.cs

### Actual Behavior
no results returned and nothing in the stack trace.  application is running, just never stops awaiting response.

### Code snippet
var results = await permissions.RequestPermissionsAsync(Permission.Location);
status = results[Permission.Location];

**Comment by jamesmontemagno** [comment]
Did you add the Permissions to your manifest?

Also did you add the code into your main activity? Did you add a break point to make sure that it gets forwarded on.

**Comment by remingtonsutton** [comment]
Yes, my permissions are:

ACCESS_COARSE_LOCATION
ACCESS_FINE_LOCATION
ACCESS_NETWORK_STATE
INTERNET
[image: Inline image 1]

yes, code is overridden in my main activity.



[image: Inline image 2]


Yes on the break point -- it simply never gets to the status.

[image: Inline image 3]

**Comment by jamesmontemagno** [comment]
Are you also compiling against Android 7.1? This is on the first page of the Android projects properties. 

So, let's walk through this based on your code:

1.) Does it  return from "CheckPermissionStatus"
2.) Does it Return from ShouldShowRequestPermissionRationale?
3.) Does it hit your breakpoint on await RequestPermissionsAsync
4.) Does it pop up a dialog box?
5.) In your MainActivity add a breakpoint on the overridden method does it hit this?

I am most interested in the 5th one to see what happens as this is waiting for the system... 

Looking at your code I am almost certain though that you will need to upgrade to using AppCompatActivity though: https://developer.xamarin.com/guides/xamarin-forms/platform-features/android/appcompat/ And this most likely is your issue.

**Comment by jamesmontemagno** [comment]
Here is your exact same code and is working great: 
[App49.zip](https://github.com/jamesmontemagno/PermissionsPlugin/files/1106995/App49.zip)


See what is different between yours and mine

**Comment by remingtonsutton** [comment]
Yes I am compiling at against android 7.1  I think I mentioned that in my
initial submission.

Yes checkpermissionstatus does work and I use this in my temporary
workaround... which is:

Plugin.Permissions.CrossPermissions.Current.RequestPermissionsAsync(new[] {
Plugin.Permissions.Abstractions.Permission.Location });

                    for (int i = 0; i < 100; i++)
                    {
                        Thread.Sleep(100);
                        status = await
Plugin.Permissions.CrossPermissions.Current.CheckPermissionStatusAsync(Plugin.Permissions.Abstractions.Permission.Location);
                        if(status ==
Plugin.Permissions.Abstractions.PermissionStatus.Granted)
                        {
                            i = 100;
                        }
                    }

Obviously, it's not ideal to do it this way.

so to go through your questions.

1.) Does it return from "CheckPermissionStatus"
yes

2.) Does it Return from ShouldShowRequestPermissionRationale?
yes, it returns false
3.) Does it hit your breakpoint on await RequestPermissionsAsync
Yes it hits the breakpoint and based on my workaround, it seems to
4.) Does it pop up a dialog box?
yes that works great, and like i said, it seems to be actually setting the
permissions correctly based on the selection of that popup dialogue
5.) In your MainActivity add a breakpoint on the overridden method does it
hit this?
NO! It isn't.  I thought about checking that a bit ago but never did.  Any
idea why??


I'm having difficulty with await Application.Current.SavePropertiesAsync();
in some cases also.  Similarly, not returning anything in SOME situations.
I'm not sure if this may be related?

I'm calling both await Application.Current.SavePropertiesAsync(); and
CheckPermissionStatusyAsync from static methods.  Just FYI.


As for AppCompat, that was a good call and I thought about that between our
last email and your response and switched it over.  It is currently
appcompat but that did NOT have any effect on the issue.

**Comment by remingtonsutton** [comment]
It IS hitting a breakpoint at RegisterActivityLifecycleCallbacks(this);
 --- so I think we can safely assume that it is being  registered.

On Tue, Jun 27, 2017 at 6:37 PM, Remington Sutton <remingtonsutton@gmail.com
> wrote:

> Yes I am compiling at against android 7.1  I think I mentioned that in my
> initial submission.
>
> Yes checkpermissionstatus does work and I use this in my temporary
> workaround... which is:
>
> Plugin.Permissions.CrossPermissions.Current.RequestPermissionsAsync(new[]
> { Plugin.Permissions.Abstractions.Permission.Location });
>
>                     for (int i = 0; i < 100; i++)
>                     {
>                         Thread.Sleep(100);
>                         status = await Plugin.Permissions.
> CrossPermissions.Current.CheckPermissionStatusAsync(Plugin.Permissions.
> Abstractions.Permission.Location);
>                         if(status == Plugin.Permissions.
> Abstractions.PermissionStatus.Granted)
>                         {
>                             i = 100;
>                         }
>                     }
>
> Obviously, it's not ideal to do it this way.
>
> so to go through your questions.
>
> 1.) Does it return from "CheckPermissionStatus"
> yes
>
> 2.) Does it Return from ShouldShowRequestPermissionRationale?
> yes, it returns false
> 3.) Does it hit your breakpoint on await RequestPermissionsAsync
> Yes it hits the breakpoint and based on my workaround, it seems to
> 4.) Does it pop up a dialog box?
> yes that works great, and like i said, it seems to be actually setting the
> permissions correctly based on the selection of that popup dialogue
> 5.) In your MainActivity add a breakpoint on the overridden method does it
> hit this?
> NO! It isn't.  I thought about checking that a bit ago but never did.  Any
> idea why??
>
>
> I'm having difficulty with await Application.Current.SavePropertiesAsync();
> in some cases also.  Similarly, not returning anything in SOME situations.
> I'm not sure if this may be related?
>
> I'm calling both await Application.Current.SavePropertiesAsync(); and
> CheckPermissionStatusyAsync from static methods.  Just FYI.
>
>
> As for AppCompat, that was a good call and I thought about that between
> our last email and your response and switched it over.  It is currently
> appcompat but that did NOT have any effect on the issue.
>
>
> On Tue, Jun 27, 2017 at 6:08 PM, James Montemagno <

**Comment by remingtonsutton** [comment]
My code is AppCompat and uses Xamarin.Forms so it's not really like the example you gave much at all.

I think the issue is in point number 5 of not hitting the overridden method -- so on that, do you have any clue as to why it is ignoring that??

**Comment by jamesmontemagno** [comment]
Based on the code you gave me it is not appcompat: https://user-images.githubusercontent.com/10901237/27601608-f39d679e-5b3d-11e7-9d51-f97a64823aef.png

Which is why I changed it. I use this exact code in appcompat and non-appcompat in about 10+ apps so I know it works and used by tons.

You could just change it to this:

```

    [Activity(Label = "App49", Icon = "@drawable/icon", Theme="@style/MainTheme", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation)]
    public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
    {
        protected override void OnCreate(Bundle bundle)
        {
            TabLayoutResource = Resource.Layout.Tabbar;
            ToolbarResource = Resource.Layout.Toolbar;

            base.OnCreate(bundle);

            global::Xamarin.Forms.Forms.Init(this, bundle);
            LoadApplication(new App());
        }

        public override void OnRequestPermissionsResult(int requestCode, string[] permissions, [GeneratedEnum] Permission[] grantResults)
        {
            Plugin.Permissions.PermissionsImplementation.Current.OnRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }
```

and it will also work just fine as I tested it.

I would highly recommend not using your Thread.Sleep as it is not needed at all as I showed in my example. 

**In summary though**
You will have to send me a sample of it not working and I am sure there is something wrong in your setup of code.

**Comment by remingtonsutton** [comment]
OK, it looks like changing the Target version to 23, 24 or 25 made absolutely no difference.  After the permissions issue I was able to get past the same geolocator issue by simply blasting the GetPositionAsync until something was returned.  It looks like it takes about 120 attempts to get something, but I can't move forward without it, so it has to get done.  I agree that these sort of loops aren't elegant, but no other solution has worked.

![image](https://user-images.githubusercontent.com/10901237/27940411-6c856240-6298-11e7-94a1-1a3884221bba.png)

**Comment by remingtonsutton** [comment]
This was a very short lived success.  It worked about 6 times before no longer working on emulator or physical device.  When it did work, it would eventually crash the application -- despite my later attempts to track and abort the threads once a location was found.

![image](https://user-images.githubusercontent.com/10901237/27941714-04cd5d94-62a0-11e7-93f7-47931b6ef66a.png)

**Comment by jamesmontemagno** [comment]
I took a look at the source code example that you sent me and unfortunately I simply do not have the time to correct or comment on all of the issues going on. 

I highly recommend looking at getting started with Async/Await Tasked based development: https://blog.xamarin.com/getting-started-with-async-await/

Here are a few tips:
1. Don't use threads like that as you have things running all over the place and locking
2. Calling .Result on a Task is going to lock your UI thread and application
3. Don't initialize and call async logic in your app constructor... it hasn't even loaded yet and you are locking the app from starting

Take a look at the sample https://github.com/jamesmontemagno/GeolocatorPlugin/tree/master/samples as it has a sample of everything you could possibly want to do with the plugin including permissions.

**Comment by remingtonsutton** [comment]
OK thanks again for taking a look at it.

I definitely didn't realize that .Result locks the UI thread.  That's good
to note, however, the issue also occurs with your example code in my
project.

I only put this code in the App() constructor so that I could get to it
faster for debugging.  The same issue occurs everywhere it's implemented.

I would have loved to not use await just like in the examples, but it
simply never returned a result.  Even when copied directly from the example
as was the case when I started this issue thread.

**Comment by remingtonsutton** [comment]
As I mentioned, this solution isn't ideal and it's a total hack.  But requesting the location continuously is the only way I ever get a response -- ever.  I know well that this isn't a proper method.  But there isn't much in terms of an explanation as to why GetPositionAsync() only returns a response every 20 requests... no matter where it's called from.

**Comment by ChaseFlorell** [comment]
```csharp
public class App
{
    private GeoCoordinate _position;
    public App()
    {
        MainPage = GetMainPage();
    }

    protected override async void OnStart()
    {
        // there's probably overhead to doing this in both OnStart and OnResume, but last time I checked OnResume also runs OnStart
        await SetPositionAsync();
    }

    protected override async void OnResume()
    {
        // there's probably overhead to doing this in both OnStart and OnResume, but last time I checked OnResume also runs OnStart
        await SetPositionAsync();
    }

    private async Task SetPositionAsync()
    {
        if(_position != null) return;
        _position = await _locator.GetPositionAsync();
    }
}
```

**Comment by imhrushi** [comment]
Is the problem solved? I am experiencing exact same problem.
In the 5 steps above, first 3 are working. However, ShouldShowRequestPermissionRationaleAsync returns true. Then I have a displayalert to show rationale, but it never shows up on screen. So I commented it out.
Then
4.) Does it pop up a dialog box?
yes RequestPermissionsAsync does pop up. When I hit deny or allow, the popup goes off, but the app hangs. The call never returns. But if I stop debugging and run again, I can see that the decision taken is saved properly. Like if I hit allow. 
5.) In your MainActivity add a breakpoint on the overridden method does it hit this?
No it does not hit this breakpoint.

Any idea how to solve this? I am checking this in VS 2017 with android oreo 8.1 API 27

**Comment by imhrushi** [comment]
by the way, the Plugin.CurrentActivity nuget package is not installing on my project. It gives error as below.
Package Plugin.CurrentActivity 2.1.0.4 is not compatible with netstandard2.0 (.NETStandard,Version=v2.0). Package Plugin.CurrentActivity 2.1.0.4 supports: monoandroid44 (MonoAndroid,Version=v4.4)

So do in need to install the monoandroid somehow first?

**Comment by wodaidai** [comment]
Hi imhrushi

I had same issue that #5 is never called. 
And found out it was due to the position of 'Plugin.CurrentActivity.CrossCurrentActivity.Current.Init' in OnCreate. I had it before base.OnCreate.

you can try to make sure:

`
        protected override void OnCreate(Bundle bundle)
        {
			FormsAppCompatActivity.ToolbarResource = Resource.Layout.Toolbar;
			FormsAppCompatActivity.TabLayoutResource = Resource.Layout.Tabbar;

			base.OnCreate(bundle);

			Plugin.CurrentActivity.CrossCurrentActivity.Current.Init(this, bundle);
            global::Xamarin.Forms.Forms.Init(this, bundle);  <--------- ** after OnCreate

            LoadApplication(new App());
        }

        public override void OnRequestPermissionsResult(int requestCode, string[] permissions, Permission[] grantResults)
        {
            PermissionsImplementation.Current.OnRequestPermissionsResult(requestCode, permissions, grantResults);
        }
`

**Comment by LasseNordqvist** [comment]
Hi get the same result. Have tested a new project from Visual studio 2019, Empty xam app only android.
Add Xam.Plugin.Media.  follow readme file. every thing works until 

` var results = await CrossPermissions.Current.RequestPermissionsAsync(Plugin.Permissions.Abstractions.Permission.Camera);`
never returns to that…  

Break point hits `public override void OnRequestPermissionsResult(int requestCode, string[] permissions, [GeneratedEnum] Android.Content.PM.Permission[] grantResults)`

Test project is added below.

[MissPer.zip](https://github.com/jamesmontemagno/PermissionsPlugin/files/3431247/MissPer.zip)

**Comment by conrad90909** [comment]
@LasseNordqvist Looking at your code you have the same issue I had.

in OnRequestPermissionsResult in MainActivity replace:

`Xamarin.Essentials.Platform.OnRequestPermissionsResult(requestCode, permissions, grantResults);`

With

`PermissionsImplementation.Current.OnRequestPermissionsResult(requestCode, permissions, grantResults);`

This solved the problem for me!

**Comment by conrad90909** [comment]
Also I don't think you want this commented out:

base.OnRequestPermissionsResult(requestCode, permissions, grantResults);

**Comment by LasseNordqvist** [comment]
> @LasseNordqvist Looking at your code you have the same issue I had.
> in OnRequestPermissionsResult in MainActivity replace:
> Xamarin.Essentials.Platform.OnRequestPermissionsResult(requestCode, permissions, grantResults);
> With
> PermissionsImplementation.Current.OnRequestPermissionsResult(requestCode, permissions, grantResults);
> This solved the problem for me!

Perfect, This solved this problem. :)

**Comment by rootflood** [comment]
ok i found the solution
we need this but this block UI thread 
``` c#
var status = Permissions.RequestAsync<Permissions.Camera>().GetAwaiter().GetResult();
```

we need other thread which can wait for result
so we change it to this and its work
``` c#
            var status = Permissions.RequestAsync<Permissions.Camera>();
            new System.Threading.Thread(() =>
            {
                status.GetAwaiter().GetResult()
            }).Start();
```
but i think it must fix because this is a logic bug

**Comment by agat366** [comment]
I am experiencing this issue. And what's strange, it's happening only on Samsung (Android 9 and 10), any other Android device seems working fine.
CheckPermissionStatusAsync works well on the fist call, but then after calling RequestPermissionAsync (and allowing it), any further calls of any of those gets to eternal waiting for result.
Any thoughts?
